ResponseErrorException when Ctrl+Hover due to cancellation

when Ctrl+Hover over qute templates or java code, several exceptions are logged, polluting the logs:

```
2023-09-01 02:02:02,258 [30716257]   WARN - c.r.d.i.l.o.n.LSPGotoDeclarationHandler - org.eclipse.lsp4j.jsonrpc.ResponseErrorException: The request (id: 120, method: 'qute/template/javaDefinition') has been cancelled
java.util.concurrent.ExecutionException: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: The request (id: 120, method: 'qute/template/javaDefinition') has been cancelled
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at com.redhat.devtools.intellij.lsp4ij.operations.navigation.LSPGotoDeclarationHandler.getGotoDeclarationTargets(LSPGotoDeclarationHandler.java:73)
	at com.intellij.codeInsight.navigation.impl.GtdProvidersKt.fromGTDProvidersInner(gtdProviders.kt:30)
	at com.intellij.codeInsight.navigation.impl.GtdProvidersKt.access$fromGTDProvidersInner(gtdProviders.kt:1)
	at com.intellij.codeInsight.navigation.impl.GtdProvidersKt$fromGTDProviders$1.invoke(gtdProviders.kt:17)
	at com.intellij.codeInsight.navigation.impl.GtdProvidersKt$fromGTDProviders$1.invoke(gtdProviders.kt)
	at com.intellij.codeInsight.navigation.impl.CommonKt.processInjectionThenHost(common.kt:25)
	at com.intellij.codeInsight.navigation.impl.GtdProvidersKt.fromGTDProviders(gtdProviders.kt:16)
	at com.intellij.codeInsight.navigation.actions.GotoDeclarationOrUsageHandler2.gotoDeclarationOrUsages(GotoDeclarationOrUsageHandler2.kt:30)
	at com.intellij.codeInsight.navigation.actions.GotoDeclarationOrUsageHandler2.getCtrlMouseData(GotoDeclarationOrUsageHandler2.kt:43)
	at com.intellij.codeInsight.navigation.actions.GotoDeclarationAction.getCtrlMouseData(GotoDeclarationAction.java:96)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2$computeInReadAction$1.invoke(CtrlMouseHandler.kt:235)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2$computeInReadAction$1.invoke(CtrlMouseHandler.kt:77)
	at com.intellij.lang.documentation.ide.impl.DocumentationTargetHoverInfoKt.injectedThenHost(DocumentationTargetHoverInfo.kt:76)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2.computeInReadAction(CtrlMouseHandler.kt:234)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2.access$computeInReadAction(CtrlMouseHandler.kt:77)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2$compute$2$1.invoke(CtrlMouseHandler.kt:222)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2$compute$2$1.invoke(CtrlMouseHandler.kt:77)
	at com.intellij.openapi.application.rw.InternalReadAction.insideReadAction(InternalReadAction.kt:96)
	at com.intellij.openapi.application.rw.InternalReadAction.access$insideReadAction(InternalReadAction.kt:13)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadCancellable$1.invoke(InternalReadAction.kt:81)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadCancellable$1.invoke(InternalReadAction.kt:13)
	at com.intellij.openapi.progress.CancellationKt$sam$com_intellij_openapi_util_ThrowableComputable$0.compute(cancellation.kt)
	at com.intellij.openapi.progress.Cancellation.withJob(Cancellation.java:60)
	at com.intellij.openapi.progress.CancellationKt.withJob(cancellation.kt:14)
	at com.intellij.openapi.progress.CancellationKt.executeWithJobAndCompleteIt(cancellation.kt:111)
	at com.intellij.openapi.application.rw.CancellableReadActionKt$cancellableReadActionInternal$2$1.run(cancellableReadAction.kt:33)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1154)
	at com.intellij.openapi.application.rw.CancellableReadActionKt$cancellableReadActionInternal$2.run(cancellableReadAction.kt:31)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:158)
	at com.intellij.openapi.application.rw.CancellableReadActionKt.cancellableReadActionInternal(cancellableReadAction.kt:29)
	at com.intellij.openapi.application.rw.InternalReadAction.tryReadCancellable(InternalReadAction.kt:80)
	at com.intellij.openapi.application.rw.InternalReadAction.access$tryReadCancellable(InternalReadAction.kt:13)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadAction$2.invoke(InternalReadAction.kt:66)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadAction$2.invoke(InternalReadAction.kt:13)
	at com.intellij.openapi.progress.CancellationKt$sam$com_intellij_openapi_util_ThrowableComputable$0.compute(cancellation.kt)
	at com.intellij.openapi.progress.Cancellation.withJob(Cancellation.java:60)
	at com.intellij.openapi.progress.CancellationKt.withJob(cancellation.kt:14)
	at com.intellij.openapi.progress.CoroutinesKt.blockingContext(coroutines.kt:138)
	at com.intellij.openapi.application.rw.InternalReadAction.tryReadAction(InternalReadAction.kt:61)
	at com.intellij.openapi.application.rw.InternalReadAction.readLoop(InternalReadAction.kt:53)
	at com.intellij.openapi.application.rw.InternalReadAction.access$readLoop(InternalReadAction.kt:13)
	at com.intellij.openapi.application.rw.InternalReadAction$runReadAction$4.invokeSuspend(InternalReadAction.kt:33)
	at com.intellij.openapi.application.rw.InternalReadAction$runReadAction$4.invoke(InternalReadAction.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:89)
	at kotlinx.coroutines.CoroutineScopeKt.coroutineScope(CoroutineScope.kt:264)
	at com.intellij.openapi.application.rw.InternalReadAction.runReadAction(InternalReadAction.kt:32)
	at com.intellij.openapi.application.rw.PlatformReadActionSupport.executeReadAction(PlatformReadActionSupport.kt:24)
	at com.intellij.openapi.application.CoroutinesKt.constrainedReadAction(coroutines.kt:47)
	at com.intellij.codeInsight.navigation.CtrlMouseHandler2$compute$2.invokeSuspend(CtrlMouseHandler.kt:221)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: The request (id: 120, method: 'qute/template/javaDefinition') has been cancelled
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
Caused by: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: The request (id: 120, method: 'qute/template/javaDefinition') has been cancelled

	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at com.redhat.devtools.intellij.lsp4ij.LanguageServerWrapper.lambda$start$1(LanguageServerWrapper.java:334)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```
 
 When hovering over java code, 'microprofile/java/definition' is being cancelled, causing the same exceptions

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationUtil.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationUtil.java
@@ -1,0 +1,49 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Avaloq Group AG.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Rubén Porras Campo (Avaloq Group AG) - Initial Implementation
+ *******************************************************************************/
+package com.redhat.devtools.intellij.lsp4ij.internal;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+
+import org.eclipse.lsp4j.jsonrpc.ResponseErrorException;
+import org.eclipse.lsp4j.jsonrpc.messages.ResponseError;
+import org.eclipse.lsp4j.jsonrpc.messages.ResponseErrorCode;
+
+/**
+ * Cancellation utilities.
+ *
+ * This class is a copy/paste from <a href="https://github.com/eclipse/lsp4e/blob/master/org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/CancellationUtil.java">https://github.com/eclipse/lsp4e/blob/master/org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/CancellationUtil.java</a>
+ */
+public final class CancellationUtil {
+
+	private CancellationUtil() {
+		// this class shouldn't be instantiated
+	}
+
+	public static boolean isRequestCancelledException(Throwable throwable) {
+		if (throwable instanceof CompletionException || throwable instanceof  ExecutionException) {
+			throwable = throwable.getCause();
+		}
+		if (throwable instanceof ResponseErrorException) {
+			return isRequestCancelled((ResponseErrorException)throwable);
+		}
+		return throwable instanceof CancellationException;
+	}
+
+	private static boolean isRequestCancelled(ResponseErrorException responseErrorException) {
+		ResponseError responseError = responseErrorException.getResponseError();
+		return responseError != null
+				&& responseError.getCode() == ResponseErrorCode.RequestCancelled.getValue();
+	}
+
+}

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPTextHoverForFile.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPTextHoverForFile.java
@@ -17,7 +17,9 @@ import com.intellij.psi.PsiElement;
 import com.redhat.devtools.intellij.lsp4ij.LSPIJUtils;
 import com.redhat.devtools.intellij.lsp4ij.LanguageServiceAccessor;
 import com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport;
+import com.redhat.devtools.intellij.lsp4ij.internal.CancellationUtil;
 import org.eclipse.lsp4j.*;
+import org.eclipse.lsp4j.jsonrpc.ResponseErrorException;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
@@ -52,15 +54,16 @@ public class LSPTextHoverForFile {
             // The LSP hover request are finished, don't need to cancel the previous LSP requests.
             previousCancellationSupport = null;
             return result;
-        } catch (ExecutionException e) {
-            if (!(e.getCause() instanceof CancellationException)) {
+        } catch (ResponseErrorException | ExecutionException | CancellationException e) {
+            // do not report error if the server has cancelled the request
+            if (!CancellationUtil.isRequestCancelledException(e)) {
                 LOGGER.warn(e.getLocalizedMessage(), e);
             }
         } catch (TimeoutException e) {
             LOGGER.warn(e.getLocalizedMessage(), e);
         } catch (InterruptedException e) {
-            LOGGER.warn(e.getLocalizedMessage(), e);
             Thread.currentThread().interrupt();
+            LOGGER.warn(e.getLocalizedMessage(), e);
         }
         return null;
     }

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/navigation/LSPGotoDeclarationHandler.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/navigation/LSPGotoDeclarationHandler.java
@@ -24,10 +24,12 @@ import com.intellij.psi.PsiManager;
 import com.redhat.devtools.intellij.lsp4ij.LSPIJUtils;
 import com.redhat.devtools.intellij.lsp4ij.LanguageServiceAccessor;
 import com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport;
+import com.redhat.devtools.intellij.lsp4ij.internal.CancellationUtil;
 import com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.ls.PsiUtilsLSImpl;
 import org.eclipse.lsp4j.DefinitionParams;
 import org.eclipse.lsp4j.Location;
 import org.eclipse.lsp4j.LocationLink;
+import org.eclipse.lsp4j.jsonrpc.ResponseErrorException;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
@@ -40,10 +42,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
 import java.util.Set;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
+import java.util.concurrent.*;
 import java.util.stream.Collectors;
 
 public class LSPGotoDeclarationHandler implements GotoDeclarationHandler {
@@ -52,33 +51,36 @@ public class LSPGotoDeclarationHandler implements GotoDeclarationHandler {
     @Nullable
     @Override
     public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement, int offset, Editor editor) {
-        try {
-            URI uri = LSPIJUtils.toUri(editor.getDocument());
-            if (uri != null) {
-                DefinitionParams params = new DefinitionParams(LSPIJUtils.toTextDocumentIdentifier(uri), LSPIJUtils.toPosition(offset, editor.getDocument()));
-                Set<PsiElement> targets = new HashSet<>();
-                final CancellationSupport cancellationSupport = new CancellationSupport();
-                try {
-                    LanguageServiceAccessor.getInstance(editor.getProject())
-                            .getLanguageServers(editor.getDocument(), capabilities -> LSPIJUtils.hasCapability(capabilities.getDefinitionProvider()))
-                            .thenComposeAsync(languageServers ->
-                                    cancellationSupport.execute(
-                                            CompletableFuture.allOf(
-                                                    languageServers
-                                                            .stream()
-                                                            .map(server ->
-                                                                    cancellationSupport.execute(server.getServer().getTextDocumentService().definition(params))
-                                                                            .thenAcceptAsync(definitions -> targets.addAll(toElements(editor.getProject(), definitions))))
-                                                            .toArray(CompletableFuture[]::new))))
-                            .get(1_000, TimeUnit.MILLISECONDS);
-                } catch (ExecutionException | TimeoutException e) {
+        URI uri = LSPIJUtils.toUri(editor.getDocument());
+        if (uri != null) {
+            DefinitionParams params = new DefinitionParams(LSPIJUtils.toTextDocumentIdentifier(uri), LSPIJUtils.toPosition(offset, editor.getDocument()));
+            Set<PsiElement> targets = new HashSet<>();
+            final CancellationSupport cancellationSupport = new CancellationSupport();
+            try {
+                LanguageServiceAccessor.getInstance(editor.getProject())
+                        .getLanguageServers(editor.getDocument(), capabilities -> LSPIJUtils.hasCapability(capabilities.getDefinitionProvider()))
+                        .thenComposeAsync(languageServers ->
+                                cancellationSupport.execute(
+                                        CompletableFuture.allOf(
+                                                languageServers
+                                                        .stream()
+                                                        .map(server ->
+                                                                cancellationSupport.execute(server.getServer().getTextDocumentService().definition(params))
+                                                                        .thenAcceptAsync(definitions -> targets.addAll(toElements(editor.getProject(), definitions))))
+                                                        .toArray(CompletableFuture[]::new))))
+                        .get(1_000, TimeUnit.MILLISECONDS);
+            } catch (ResponseErrorException | ExecutionException | CancellationException e) {
+                // do not report error if the server has cancelled the request
+                if (!CancellationUtil.isRequestCancelledException(e)) {
                     LOGGER.warn(e.getLocalizedMessage(), e);
                 }
-                return targets.toArray(new PsiElement[targets.size()]);
+            } catch (TimeoutException e) {
+                LOGGER.warn(e.getLocalizedMessage(), e);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                LOGGER.warn(e.getLocalizedMessage(), e);
             }
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            LOGGER.warn(e.getLocalizedMessage(), e);
+            return targets.toArray(new PsiElement[targets.size()]);
         }
         return new PsiElement[0];
     }

